File: stdlib/js/hull/cookie.js
/*
 * hull:cookie -- Cookie parsing and serialization
 *
 * cookie.parse(headerString)        -> object of name/value pairs
 * cookie.serialize(name, value, opts) -> Set-Cookie header string
 * cookie.clear(name, opts)          -> Set-Cookie header that expires the cookie
 *
 * Defaults: httpOnly=true, secure=false, sameSite="Lax", path="/"
 *
 * SPDX-License-Identifier: AGPL-3.0-or-later
 */

function parse(headerString) {
    const result = {};
    if (!headerString || typeof headerString !== "string")
        return result;

    const pairs = headerString.split(";");
    for (let i = 0; i < pairs.length; i++) {
        const pair = pairs[i].trim();
        if (pair.length === 0) continue;

        const eqIdx = pair.indexOf("=");
        if (eqIdx < 0) continue;

        const name = pair.substring(0, eqIdx).trim();
        if (name.length === 0) continue;

        let value = pair.substring(eqIdx + 1).trim();

        // Strip surrounding quotes if present
        if (value.length >= 2 && value[0] === '"' && value[value.length - 1] === '"')
            value = value.substring(1, value.length - 1);

        // Decode percent-encoded values
        try {
            result[name] = decodeURIComponent(value);
        } catch (e) {
            result[name] = value;
        }
    }

    return result;
}

function serialize(name, value, opts) {
    if (!name || typeof name !== "string")
        throw new Error("cookie name is required");

    const o = opts || {};

    // Percent-encode the value
    let encoded;
    if (value === null || value === undefined)
        encoded = "";
    else
        encoded = encodeURIComponent(String(value));

    let str = name + "=" + encoded;

    // Path (default: "/")
    const path = o.path !== undefined ? o.path : "/";
    if (path)
        str += "; Path=" + path;

    // Domain
    if (o.domain)
        str += "; Domain=" + o.domain;

    // MaxAge
    if (o.maxAge !== undefined && o.maxAge !== null) {
        const maxAge = Math.floor(o.maxAge);
        if (isNaN(maxAge))
            throw new Error("maxAge must be a number");
        str += "; Max-Age=" + maxAge;
    }

    // Expires
    if (o.expires) {
        if (typeof o.expires === "string")
            str += "; Expires=" + o.expires;
        else if (typeof o.expires === "number")
            str += "; Expires=" + new Date(o.expires * 1000).toUTCString();
    }

    // HttpOnly (default: true)
    const httpOnly = o.httpOnly !== undefined ? o.httpOnly : true;
    if (httpOnly)
        str += "; HttpOnly";

    // WARNING: Set secure=true in production (HTTPS). Default false for local dev only.
    const secure = o.secure !== undefined ? o.secure : false;
    if (secure)
        str += "; Secure";

    // SameSite (default: "Lax")
    const sameSite = o.sameSite !== undefined ? o.sameSite : "Lax";
    if (sameSite) {
        const ss = String(sameSite);
        if (ss === "Strict" || ss === "Lax" || ss === "None")
            str += "; SameSite=" + ss;
        else
            throw new Error("sameSite must be Strict, Lax, or None");
    }

    return str;
}

function clear(name, opts) {
    const o = Object.assign({}, opts || {});
    o.maxAge = 0;
    o.expires = "Thu, 01 Jan 1970 00:00:00 GMT";
    return serialize(name, "", o);
}

const cookie = { parse, serialize, clear };
export { cookie };

File: stdlib/lua/hull/cookie.lua
--
-- hull.cookie -- Cookie parsing and serialization
--
-- Pure functions for HTTP cookie handling.
--
-- SPDX-License-Identifier: AGPL-3.0-or-later
--

local cookie = {}

--- Parse a Cookie header string into a key-value table.
-- "session=abc; theme=dark" -> { session = "abc", theme = "dark" }
function cookie.parse(header_string)
    local result = {}
    if not header_string or header_string == "" then
        return result
    end

    for pair in string.gmatch(header_string, "[^;]+") do
        -- Trim leading/trailing whitespace
        pair = pair:match("^%s*(.-)%s*$")
        if pair ~= "" then
            local eq = pair:find("=", 1, true)
            if eq then
                local name = pair:sub(1, eq - 1):match("^%s*(.-)%s*$")
                local value = pair:sub(eq + 1):match("^%s*(.-)%s*$")
                if name ~= "" then
                    result[name] = value
                end
            end
        end
    end

    return result
end

--- Serialize a cookie name/value pair with options into a Set-Cookie header value.
-- Defaults: HttpOnly=true, Secure=false, SameSite=Lax, Path=/
function cookie.serialize(name, value, opts)
    opts = opts or {}

    local parts = { name .. "=" .. (value or "") }

    -- Path (default "/")
    local path = opts.path
    if path == nil then path = "/" end
    if path then
        parts[#parts + 1] = "Path=" .. path
    end

    -- HttpOnly (default true)
    local httponly = opts.httponly
    if httponly == nil then httponly = true end
    if httponly then
        parts[#parts + 1] = "HttpOnly"
    end

    -- WARNING: Set secure=true in production (HTTPS). Default false for local dev only.
    local secure = opts.secure
    if secure == nil then secure = false end
    if secure then
        parts[#parts + 1] = "Secure"
    end

    -- SameSite (default "Lax")
    local samesite = opts.samesite
    if samesite == nil then samesite = "Lax" end
    if samesite then
        parts[#parts + 1] = "SameSite=" .. samesite
    end

    -- Max-Age
    if opts.max_age then
        parts[#parts + 1] = "Max-Age=" .. tostring(opts.max_age)
    end

    -- Domain
    if opts.domain then
        parts[#parts + 1] = "Domain=" .. opts.domain
    end

    -- Expires
    if opts.expires then
        parts[#parts + 1] = "Expires=" .. opts.expires
    end

    return table.concat(parts, "; ")
end

--- Clear a cookie by setting Max-Age=0 and an empty value.
function cookie.clear(name, opts)
    opts = opts or {}
    local clear_opts = {}
    -- Copy relevant options
    clear_opts.path = opts.path
    clear_opts.domain = opts.domain
    clear_opts.httponly = opts.httponly
    clear_opts.secure = opts.secure
    clear_opts.samesite = opts.samesite
    clear_opts.max_age = 0
    return cookie.serialize(name, "", clear_opts)
end

return cookie
